driver-core/src/main/com/mongodb/connection/ScramShaAuthenticator.java

```diff
@@ -373,4 +384,64 @@ public String generate(final MongoCredential credential) {
     private static AuthenticationHashGenerator getAuthenicationHashGenerator(final AuthenticationMechanism authenticationMechanism) {
         return authenticationMechanism == SCRAM_SHA_1 ? LEGACY_AUTHENTICATION_HASH_GENERATOR : DEFAULT_AUTHENTICATION_HASH_GENERATOR;
     }
+
+    private static Map<CacheKey, CacheValue> createKeyCache() {
+        return Collections.synchronizedMap(new LinkedHashMap<CacheKey, CacheValue>() {
+            private static final long serialVersionUID = 1L;
+            @Override
+            protected boolean removeEldestEntry(final Map.Entry<CacheKey, CacheValue> eldest) {
+                return size() >= KEY_CACHE_SIZE;
+            }
+        });
+    }
+
+    private static class CacheKey {
+        private final String password;
```

Comment:
Let's call this hashedPasswordAndSalt so it's clear what it actually is.

Reply:
Done.